File: src/main/java/org/sinnlabs/ui/zk/bootstrap/Searchmenu.java
/**
 *  {{IS_RIGHT
 *  This program is distributed under LGPL Version 2.1 in the hope that
 *  it will be useful, but WITHOUT ANY WARRANTY.
 *  }}IS_RIGHT
 */
package org.sinnlabs.ui.zk.bootstrap;

import java.util.Objects;

import org.zkoss.zk.ui.Component;
import org.zkoss.zk.ui.UiException;
import org.zkoss.zul.Menu;
import org.zkoss.zul.Menuitem;
import org.zkoss.zul.Menupopup;
import org.zkoss.zul.Menuseparator;
import org.zkoss.zul.Textbox;

/**
 * Represents popup menu with filtering
 * @author nsa_a1
 * @since 0.0.3
 */
public class Searchmenu extends Menupopup {

	/**
	 * 
	 */
	private static final long serialVersionUID = -6834729886288546928L;
	
	/**
	 * Minimum count of items to display filter
	 */
	protected int minItemCount_ = 10;
	
	protected String placeHolder_ = "Search...";
	
	public Searchmenu() {
		super();
	}
	
	/**
	 * Minimum count of items to display filter
	 * <strong>Default:</strong> 10
	 * @return
	 */
	public int getMinItemCount() {
		return minItemCount_;
	}
	
	/**
	 * Sets minium item count to display filter
	 * @param count
	 */
	public void setMinItemCount(int count) {
		if (minItemCount_ != count) {
			minItemCount_ = count;
			smartUpdate("minItemCount", minItemCount_);
		}
	}
	
	/**
	 * Returns search bar place holder text
	 * <strong>Default</strong> Search...
	 * @return
	 */
	public String getPlaceHolder() {
		return placeHolder_;
	}
	
	/**
	 * Sets search bar placeholder
	 * @param val placeholder text
	 */
	public void setPlaceHolder(String val) {
		if (!Objects.equals(val, placeHolder_)) {
			placeHolder_ = val;
			smartUpdate("placeHolder", placeHolder_);
		}
	}
	
	@Override
	protected void renderProperties(org.zkoss.zk.ui.sys.ContentRenderer renderer)
			throws java.io.IOException {
		super.renderProperties(renderer);

		render(renderer, "minItemCount", minItemCount_);
		render(renderer, "placeHolder", placeHolder_);
	}

}

File: src/main/resources/web/js/bootstrap/Searchmenu.js
/*

	Copyright (c) 2019 Sinnlabs LTD, All Rights Reserved.

	This program is distributed under LGPL Version 2.1 in the hope that
	it will be useful, but WITHOUT ANY WARRANTY.

	author: nsa_a1
 */

bootstrap.Searchmenu = zk.$extends(zul.menu.Menupopup, {
	_minItemCount : 10,  /* represents minimum items count to show search bar */
	
	_placeHolder : 'Search...',

	$define: {
		minItemCount : function(val) {
			if (val<0)
				val = 0;
			this._minItemCount = val;
		},
		placeHolder : function(val) {
			this._placeholder = val;
		}
	},

	// override some default functions
	_inBSMold: function () {
		return this._mold == 'bootstrap';
	},
	// Replace zclass
	getZclass: function () {
		if (this._inBSMold()) {
			return this._zclass ? this._zclass : '';
		} else
			return this.$super('getZclass', arguments);
	},
	// Replace classes
	$s: function (subclass) {
		if (this._inBSMold()) {
			switch (subclass) {
			case 'separator':
				return '';
			case 'content':
				return 'dropdown-menu';
			}
			return '';
		} else
			return this.$super('$s', arguments);
	},
	// Override open to support bootstrap
	open: function (ref, offset, position, opts) {
		if (this._inBSMold()) {
			this.forcerender();
			jq(this.$n('cave')).css({position: 'relative', display: 'block'});
			this.$super('open', ref, offset, position, opts);
			if (this.$n('filter')) {
				jq(this.$n('filter')).focus();
			}
		} else 
			return this.$super('open', ref, offset, position, opts);
	},
	
	doKeyDown_ : function(evt) {
		if (this.desktop) {
			jq(this.$n('filter')).trigger({type: 'keypress', which: evt.keyCode, keyCode: evt.keyCode});
		}
	},
	
	doKeyUp_: function (evt) {
		if (this.desktop && this.$n('filter')) {
			var value = jq(this.$n('filter')).val().toLowerCase();
			jq('#' + this.uuid + '-cave li:not(:first-child)' ).filter(function() {
			      $(this).toggle($(this).text().toLowerCase().indexOf(value) > -1);
		    });;
		}
	}
})